File: RPNCalculator.hs
-- 逆ポーランド記法電卓 (関数型スタイル)

module RPNCalculator (
  calculate,
  main
) where

import Data.Char (isDigit)

-- | 逆ポーランド記法の文字列を読み込み、計算する。
--
-- >>> calculate "1 2 + 3 / 4 - 5 *"
-- -15.0
-- >>> calculate "1 2 + 3 & 4 - 5 *"
-- *** Exception: unsupported operator '&' is used
-- >>> calculate "1 2 + / 3 4 - 5 *"
-- *** Exception: unexpected pattern found
-- >>> calculate ""
-- *** Exception: empty input
calculate :: String -> Double
calculate rpn = case expr of
  [] -> error "empty input"
  _  -> calc [] expr
  where
    calc stack [] = head stack
    calc stack (x:xs)
      | all isDigit x = calc ((read x :: Double) : stack) xs
      | otherwise     =
        case stack of
          (y1:y2:ys) -> calc (opeFn x y2 y1 : ys) xs
          _          -> error "unexpected pattern found"
    expr = words rpn

opeFn :: Fractional a => String -> a -> a -> a
opeFn "+" = (+)
opeFn "-" = (-)
opeFn "*" = (*)
opeFn "/" = (/)
opeFn ope = error $ "unsupported operator '" ++ ope ++ "' is used"

-- 利用例
main :: IO ()
main = putStrLn $ rpn ++ "\n = " ++ show (calculate rpn)
  where
    rpn = "1 2 + 3 / 4 - 5 *"

File: rpn_calculator.erl
%% 逆ポーランド記法電卓 (関数型スタイル)

-module(rpn_calculator).
-export([calculate/1]).
-export([main/0]).

calculate(Rpn) ->
  Calc = fun
    Calc(Stack, []) ->
      hd(Stack);
    Calc(Stack, [X|Xs]) ->
      case re:run(X, "^[0-9]+$") of
        {match, _} ->
          Calc([list_to_integer(X) | Stack], Xs);
        _ ->
          case Stack of
            [Y1,Y2|Ys] ->
              OpeFn = ope_fn(X),
              Calc([OpeFn(Y2, Y1) | Ys], Xs)
          end
      end
  end,
  Expr = string:tokens(Rpn, " "),
  Calc([], Expr).

ope_fn("+") ->
  fun(X, Y) -> X + Y end;
ope_fn("-") ->
  fun(X, Y) -> X - Y end;
ope_fn("*") ->
  fun(X, Y) -> X * Y end;
ope_fn("/") ->
  fun(X, Y) -> X / Y end;
ope_fn(Ope) ->
  error("unsupported operator '" ++ Ope ++ "' is used").

%% 利用例
main() ->
  Rpn = "1 2 + 3 / 4 - 5 *",
  io:format("~s~n = ~p~n", [Rpn, calculate(Rpn)]).
